### PR TITLE
Core: Local cache for video bids

### PR DESCRIPTION
**Pull request by mkomorski**
## Type of change

- [x] Feature

### CODE REVIEW

File: src/videoCache.js [hunk 1]
```diff
@@ -162,6 +170,36 @@ export function getCacheUrl(id) {
...
 
+export const storeLocally = (bid) => {
+  const vastValue = getVastValue(bid);
+  const dataUri = 'data:text/xml;base64,' + btoa(vastValue);
+  bid.vastUrl = dataUri;
+  //@todo: think of wrapping it with [if (adServer)]
+  vastsLocalCache.set(getLocalCacheBidId(bid), dataUri);
+};
+
+export async function getLocalCachedBidWithGam(adTagUrl) {
```
Comment (dgirardi) [inline src/videoCache.js:181]: This should be in the dfp module. 

IMO the GAM flow in general needs to change, instead of calling dfp's buildVideoUrl and passing the URL to the player it should be something like getVast and pass the XML to the player, which should work regardless of whether one is using local cache, and should also be the same thing that's called by the video module (regardless of the type of cache in use).

It might help to update / add a non-video module example to flesh out how the new API should look.

File: src/videoCache.js [hunk 2]
```diff
@@ -162,6 +170,36 @@ export function getCacheUrl(id) {
...
+  const response = await fetch(gamAdTagUrl);
+
+  if (!response.ok) {
+    logError('Unable to fetch valid response from Google Ad Manager');
+    return;
+  }
+
+  const gamVastWrapper = await response.text();
+  const bidVastDataUri = vastsLocalCache.get(`${hb_bidder}_${hb_adid}`);
+  const mockUrl = LOCAL_CACHE_MOCK_URL + hb_bidder;
```
Comment (dgirardi) [inline src/videoCache.js:195]: the "normal" cache generates a UUID which is then passed on to the creative and echoed back in the VAST.

I think the local cache could also generate a UUID, pass it along in the same way, and this could look for it in the response contents instead of imposing a different URL structure in the creative. It'd have the advantage of working with existing integrations, and being easier to switch into / out of.

**Comment by pm-azhar-mulla** [comment]
We also need to update the functionality that triggers the **BID_WON** event, particularly for instream video scenarios handled within the **instreamTracking** module.

File: modules/dfpAdServerVideo.js [hunk 3]
```diff
@@ -249,6 +253,55 @@ function getCustParams(bid, options, urlCustParams) {
   return encodedParams;
 }
 
+export async function getBidVastWithGam(adTagUrl, cacheMap = vastsLocalCache) {
+  const gamAdTagUrl = new URL(adTagUrl);
+  const custParams = new URLSearchParams(gamAdTagUrl.searchParams.get('cust_params'));
+  const videoCacheKey = custParams.get('hb_uuid');
```
Comment (dgirardi) [inline modules/dfpAdServerVideo.js:259]: Id prefer not doing the URL parsing, since it couples this to the URL structure which is also tied to the ad server. Since the local cache generates the UUIDs, it seems like the response parser could look for them without first checking the URL.

File: modules/videojsVideoProvider.js [hunk 4]
```diff
@@ -201,13 +201,17 @@ export function VideojsProvider(providerConfig, vjs_, adState_, timeState_, call
 
   // Plugins to integrate: https://github.com/googleads/videojs-ima
   function setAdTagUrl(adTagUrl, options) {
```
Comment (dgirardi) [inline modules/videojsVideoProvider.js:203]: It'd be cleaner if we had a separate `setAdXml` rather than a `setAdTagUrl` that doesn't actually set the ad tag url. (same with jwplayer)

File: src/videoCache.js [hunk 5]
```diff
@@ -162,6 +175,20 @@ export function getCacheUrl(id) {
   return `${config.getConfig('cache.url')}?uuid=${id}`;
 }
 
+export const storeLocally = (bid) => {
+  const vastValue = getVastXml(bid);
+  const strategy = config.getConfig('cache.strategy') || BLOB;
```
Comment (dgirardi) [inline src/videoCache.js:180]: Why expose these / make them mutually exclusive? I'd always generate a blob and translate it to a string only if/when we need to. It's less for the pub to understand and it also allows using different players with the same bid.

File: src/videoCache.js [hunk 6]
```diff
@@ -162,6 +175,20 @@ export function getCacheUrl(id) {
...
+  const bidVastUrl = {
+    [DATA]: (xml) => 'data:text/xml;base64,' + btoa(xml),
+    [BLOB]: (xml) => URL.createObjectURL(new Blob([xml], { type: 'text/xml' }))
+  }[strategy](vastValue);
+
+  bid.vastUrl = bidVastUrl;
+  bid.videoCacheKey = generateUUID();
```
Comment (dgirardi) [inline src/videoCache.js:187]: this should be shared with the "normal" cache flow - so that they don't get out of sync if the logic changes

File: libraries/video/shared/vastXmlEditor.js [hunk 7]
```diff
@@ -74,6 +78,46 @@ export function VastXmlEditor(xmlUtil_) {
...
+      const vastAdTagUriElements = xmlDoc.querySelectorAll(VAST_TAG_URI_TAGNAME);
+      const vastAdTagUriElement = Array.from(vastAdTagUriElements).find(adTag => adTag.textContent.includes(match));
+
+      if (vastAdTagUriElement) {
+        const cdata = xmlDoc.createCDATASection(newContent);
+        vastAdTagUriElement.textContent = '';
+        vastAdTagUriElement.appendChild(cdata);
+        return xmlUtil.serialize(xmlDoc);
+      } else {
+        throw new Error();
```
Comment (dgirardi) [inline libraries/video/shared/vastXmlEditor.js:94]: this generates an error with no reason

File: modules/dfpAdServerVideo.js [hunk 8]
```diff
@@ -22,6 +22,9 @@ import {
   parseUrl
 } from '../src/utils.js';
 import {DEFAULT_DFP_PARAMS, DFP_ENDPOINT, gdprParams} from '../libraries/dfpUtils/dfpUtils.js';
+import { vastsLocalCache } from '../src/videoCache.js';
+import { vastXmlEditorFactory } from '../libraries/video/shared/vastXmlEditor.js';
```
Comment (dgirardi) [inline modules/dfpAdServerVideo.js:26]: this now includes the video library with dfpAdServerVideo. It's better than in core, but still way more than you need. It's best to extract just the bit you need.

File: modules/videoModule/index.js [hunk 9]
```diff
@@ -84,11 +85,26 @@ export function PbVideo(videoCore_, getConfig_, pbGlobal_, pbEvents_, videoEvent
...
+      if (!response.ok) {
+        throw new Error('VAST fetch failed');
+      }
+
+      let vastXml = await response.text();
+
+      // If the fetched VAST XML is a GAM wrapper and local cached is being used,
+      // retrieve the bid's VAST blob from the local cache
+      // and replace the VastAdTagURI with its content.
+      if (config.getConfig('cache.useLocal') && vastXml.includes(bid.videoCacheKey)) {
```
Comment (dgirardi) [inline modules/videoModule/index.js:100]: This logic is duplicated with what's in the dfp module. You can only get here from `renderWinningBid`, so see below.

File: modules/videoModule/index.js [hunk 10]
```diff
@@ -206,38 +222,35 @@ export function PbVideo(videoCore_, getConfig_, pbGlobal_, pbEvents_, videoEvent
...
     const videoConfig = adUnit.video;
     const divId = videoConfig.divId;
+
     const adServerConfig = getAdServerConfig(videoConfig);
-    let adUrl;
-    if (adServerConfig) {
-      adUrl = gamSubmodule.getAdTagUrl(adUnit, adServerConfig.baseAdTagUrl, adServerConfig.params);
-    }
+    const winningBid = getWinningBid(adUnitCode);
+    const vastUrl = !adServerConfig ? winningBid?.vastUrl : gamSubmodule.getAdTagUrl(
```
Comment (dgirardi) [inline modules/videoModule/index.js:231]: It seems to me all the changes to this file could instead be:
 - call `gamSubmodule.getVast` instead of `gamSubmodule.getAdTagUrl`
 - make sure the vast XML and not the URL is used.

File: modules/videojsVideoProvider.js [hunk 11]
```diff
@@ -216,6 +216,22 @@ export function VideojsProvider(providerConfig, vjs_, adState_, timeState_, call
...
+      return;
+    }
+
+    // The VideoJS IMA plugin version 1.11.0 will throw when the ad is empty.
+    try {
+      player.ima.controller.settings.adsResponse = vastXml;
+      player.ima.requestAds();
+    } catch (e) {
+      /*
+      Handling is not required; ad errors are emitted automatically by video.js
```
Comment (dgirardi) [inline modules/videojsVideoProvider.js:230]: does this sound right @karimMourra ?

File: modules/videoModule/index.js [hunk 12]
```diff
@@ -88,7 +89,15 @@ export function PbVideo(videoCore_, getConfig_, pbGlobal_, pbEvents_, videoEvent
     const adUrl = bid.vastUrl;
     options.adXml = bid.vastXml;
     options.winner = bid.bidder;
-    loadAdTag(adUrl, divId, options);
+
+    if (options.adXml) {
+      loadAd(options.adXml, divId, options);
+    } else {
+      fetch(adUrl)
```
Comment (karimMourra) [inline modules/videoModule/index.js:96]: I don't know if it's a good idea to fetch the ad tag URL when it is not a blob. For example, the IMA SDK will append query params to the ad tag URL when making requests. In this case the request to the ad server would be different, potentially resulting in lower performance. I think this fetch should only happen if the URL is a blob. If it's an external URL it should be passed to the player.

File: modules/videoModule/index.js [hunk 13]
```diff
@@ -204,40 +213,35 @@ export function PbVideo(videoCore_, getConfig_, pbGlobal_, pbEvents_, videoEvent
...
-
-    renderBid(divId, highestBid, options);
+    return highestCpmBids.shift();
   }
 
-  // options: adXml, winner, adUnitCode,
-  function loadAdTag(adTagUrl, divId, options) {
-    adQueueCoordinator.queueAd(adTagUrl, divId, options);
+  function loadAd(adXml, divId, options) {
+    adQueueCoordinator.queueAd(adXml, divId, options);
```
Comment (karimMourra) [inline modules/videoModule/index.js:244]: loadAd should not only load Ad XML, i don't think this is realistic.

File: modules/videoModule/adQueue.js [hunk 14]
```diff
@@ -9,13 +9,13 @@ export function AdQueueCoordinator(videoCore, pbEvents) {
     videoCore.onEvents([SETUP_COMPLETE], onSetupComplete, divId);
   }
 
-  function queueAd(adTagUrl, divId, options) {
+  function queueAd(adXml, divId, options) {
     const queue = storage[divId];
     if (queue) {
-      queue.push({adTagUrl, options});
-      triggerEvent(AUCTION_AD_LOAD_QUEUED, adTagUrl, options);
```
Comment (karimMourra) [inline modules/videoModule/adQueue.js:16]: this is breaking the event signature; the param is `adTagUrl` when the queued ad is an ad tag URL. Clients could be  expecting this value.  Why is this Queueing system being changed to only work with Ad XML ?

File: modules/dfpAdServerVideo.js [hunk 15]
```diff
@@ -249,6 +254,74 @@ function getCustParams(bid, options, urlCustParams) {
...
+  const blobUrl = vastsLocalCache.get(bid.videoCacheKey);
+  if (!blobUrl) return gamVastWrapper;
+  const finalVast = await replaceVastAdTagWithBlobContent(gamVastWrapper, blobUrl, bid.videoCacheKey);
+  return finalVast;
+};
+
+async function getVast(options) {
+  const { bid } = options;
+  const vastUrl = buildDfpVideoUrl(options);
+  const response = await fetch(vastUrl);
```
Comment (karimMourra) [inline modules/dfpAdServerVideo.js:267]: I would advise against this; the IMA SDK should be requesting GAM URLs. THe IMA SDK appends query params so my understanding is you would get different responses. Unwrapping the VAST should be opt-in

**Review by karimMourra — CHANGES_REQUESTED** [review]
We should not entirely change the video module to only work with VAST XML; unfortunately in some cases its best for the IMA SDK to request the Ad Tag URL.

File: modules/videoModule/index.js (hunk 12 above)
Comment (dgirardi) [reply modules/videoModule/index.js:96]: The issue is that the point of using a cache (and a local cache) is to have something that GAM can echo back if Prebid wins. So in practice this is never a blob but a dfp URL that may point to a locally cached vast.  GAM cannot deal with blobs, hence the fetch and swap here.


I was not aware that ima would mess the request but in retrospect that's not surprising. A couple questions:

 - do you know if it's possible to get just the parameters from IMA?
 - is IMA different from videojs? could we do this for the video module + jwplayer for example.

Comment (karimMourra) [reply modules/videoModule/index.js:96]: I don't think this solution works then. Has this been tested ? Unless things have changed, my understanding of the flow is:
- Prebid auction results in a winning bidder
- winning bidder ad is cached in prebid cache
- GAM ad tag URL is modified to include cache URL and cache ID in query params, as well as bidder CMP etc
- GAM ad tag URL is loaded in IMA SDK
- If the Prebid Winner also wins the GAM Auction, then one of the VAST Wrappers points to the prebid cache

In other words, GAM is loading all the VAST wrappers in the chain, including the one in the local cache. As far as I understand, there is no opportunity to intercept the request.

Comment (karimMourra) [reply modules/videoModule/index.js:96]: > do you know if it's possible to get just the parameters from IMA?
I don't think so; IMA (google) try to act as a black box; even if we reverse engineer the params that are appended, new params may be added in the future
> is IMA different from videojs? could we do this for the video module + jwplayer for example.
Video.js uses an IMA plugin which is a wrapper around the IMA SDK. There are other ad plugins, but the general consensus is that the best ad plugin to use for GAM urls is the IMA SDK.

Comment (karimMourra) [reply modules/videoModule/index.js:96]: This PR is changing the way the Video Module works with GAM, because it's unwrapping all VAST URLs before sending to the player, and I think that could result in a loss in performance.
What I do suggest is either only making these changes when the ad is cached locally, or making it opt-in for the publisher

Comment (dgirardi) [reply modules/videoModule/index.js:96]: >      In other words, GAM is loading all the VAST wrappers in the chain

No, GAM just wraps the URL you provide to it with its trackers. This works if you take IMA out.

>    Video.js uses an IMA plugin which is a wrapper around the IMA SDK. There are other ad plugins, but the general consensus is that the best ad plugin to use for GAM urls is the IMA SDK.

Ok, but what's the relationship between the video module and IMA when not using videojs? If there's none, this wouldn't be a problem except for videojs, right?

Comment (karimMourra) [reply modules/videoModule/index.js:96]: > No, GAM just wraps the URL you provide to it with its trackers. This works if you take IMA out.

I'm not sure what you mean by this. What i meant to say is *the IMA SDK* is loading all the VAST wrappers in the chain. 

> Ok, but what's the relationship between the video module and IMA when not using videojs? If there's none, this wouldn't be a problem except for videojs, right?

JW Player has multiple ad plugins, one of which is an IMA plugin which uses the IMA SDK. It is the most used ad plugin, and all publishers using GAM Ad Server use it. I suspect most commercial players use it.

Comment (karimMourra) [reply modules/videoModule/index.js:96]: Another drawback with these changes is it forces all video players to have a `loadAdXml` equivalent; that may not be the case for all players. The norm is to support a function to load ad tag URLs

Comment (karimMourra) [reply modules/videoModule/index.js:98]: Here's how prebid.js works with ad servers for 2nd auctions:

In [Keyword targeting for all bidders](https://docs.prebid.org/dev-docs/publisher-api-reference/bidderSettings.html)  

> video will receive additional keys: hb_cache_id and hb_cache_host. The key value pair targeting is applied to the bid’s corresponding ad unit. Your ad ops team will have the ad server’s line items and creatives to utilize these keys.

In [In Player Video Cache Location](https://docs.prebid.org/adops/setting-up-prebid-video-in-dfp.html)

> If you’re using a Send Top Price Bid configuration, then the VAST URL will be the same for each bidder:
https://prebid.adnxs.com/pbc/v1/cache?uuid=%%PATTERN:hb_uuid%%
or
[other bidder cache location]
If you’re using Send All Bids, the VAST URL will include the bidder-specific targeting variable. Be sure to replace BIDDERCODE with the actual bidder code for your bidders:
https://prebid.adnxs.com/pbc/v1/cache?uuid=%%PATTERN:hb_uuid_BIDDERCODE%%
or
[other bidder cache location]

**Review by karimMourra — CHANGES_REQUESTED** [review]

File: modules/dfpAdServerVideo.js [hunk 16]
```diff
@@ -249,6 +254,68 @@ function getCustParams(bid, options, urlCustParams) {
...
+  try {
+    const xmlUtil = XMLUtil();
+    const xmlDoc = xmlUtil.parse(gamVastWrapper);
+    const vastAdTagUriElement = xmlDoc.querySelectorAll(VAST_TAG_URI_TAGNAME)[0];
+
+    if (!vastAdTagUriElement || !vastAdTagUriElement.textContent) {
+      return gamVastWrapper;
+    }
+
+    const uuidExp = new RegExp(`.*(?:([A-Fa-f0-9]{8}-(?:[A-Fa-f0-9]{4}-){3}[A-Fa-f0-9]{12}).*)+`, 'gi');
```
Comment (dgirardi) [inline modules/dfpAdServerVideo.js:267]: `[A-Fa-f0-9]{8}-(?:[A-Fa-f0-9]{4}-){3}[A-Fa-f0-9]{12}` should work and with `matchAll` give you all UUID candidates (since you're also giving it the 'g' global flag). if you match everything else (wrapping it in `.*(...).*`) it will match only once and give you the one capturing group within.

File: modules/dfpAdServerVideo.js [hunk 17]
```diff
@@ -249,6 +254,68 @@ function getCustParams(bid, options, urlCustParams) {
...
+    if (uuid) {
+      const blobUrl = localCacheMap.get(uuid);
+      const base64BlobContent = await getBase64BlobContent(blobUrl);
+      const cdata = xmlDoc.createCDATASection(base64BlobContent);
+      vastAdTagUriElement.textContent = '';
+      vastAdTagUriElement.appendChild(cdata);
+      return xmlUtil.serialize(xmlDoc);
+    } else {
+      throw new Error('Unable to find uuid in ad tag');
+    }
```
Comment (dgirardi) [inline modules/dfpAdServerVideo.js:280]: This can just check if the uuid is in the map no? no need to iterate over it. And when it's not in the map, or when it finds multiple that are in in it, it should return the unmodified VAST - not throw an error.

File: modules/videoModule/index.js [hunk 18]
```diff
@@ -88,7 +89,20 @@ export function PbVideo(videoCore_, getConfig_, pbGlobal_, pbEvents_, videoEvent
...
+    // in case of using local cache, VAST should be prefetched
+    // needed for make blob always reachable, regardless of player context
+    if (useLocal && !options.prefetchedVastXml) {
+      fetch(adUrl)
+        .then(response => response.text())
+        .then(prefetchedVastXml => loadAd(adUrl, divId, {...options, prefetchedVastXml}))
+        .catch((err) => logError(`Unable to fetch vast for bid ${bid.bidder}`, err))
+      return;
+    }
+
```
Comment (karimMourra) [inline modules/videoModule/index.js:104]: Why are these changes necessary ? I don't think this is needed. This function is to render a given Bid in a video player, it has nothing to do with caching. The cache is never used here, why fetch the vast xml ?

**Review by karimMourra — CHANGES_REQUESTED** [review]

File: test/spec/modules/dfpAdServerVideo_spec.js [hunk 19]
```diff
@@ -706,4 +709,167 @@ describe('The DFP video support module', function () {
...
+
+    server.respondWith(/^https:\/\/pubads.*/, gamWrapper);
+    server.respondWith(/^blob:http:*/, blobContent);
+
+    getVastXml({url, adUnit: {}, bid: {}}, localMap)
+      .then((vastXml) => {
+        expect(vastXml).to.deep.eql(expectedOutput);
+        done();
+      })
+      .catch(done)
```
Comment (karimMourra) [inline test/spec/modules/dfpAdServerVideo_spec.js:775]: why is this line necessary ? Wouldn't we want the test to fail if an exception is thrown ?

File: test/spec/modules/dfpAdServerVideo_spec.js [hunk 20]
```diff
@@ -706,4 +709,167 @@ describe('The DFP video support module', function () {
...
+       `</Ad>` +
+      `</VAST>`
+    );
+    server.respondWith(gamWrapper);
+    getVastXml({}, null) // exception thrown when passing null as localCacheMap
+      .then((finalGamWrapper) => {
+        expect(finalGamWrapper).to.deep.eql(gamWrapper);
+        done();
+      })
+      .catch(done)
```
Comment (karimMourra) [inline test/spec/modules/dfpAdServerVideo_spec.js:871]: why is done called twice ?

**Review by karimMourra — CHANGES_REQUESTED** [review]

File: modules/videoModule/index.js (hunk 18 above)
Comment (karimMourra) [reply modules/videoModule/index.js:104]: So what you're saying is even if there is no GAM or Ad Server, we always cache the bid locally as long as `cache.useLocal` is defined ? In that case I think this flow makes sense, though in practice this probably won't be a common use case since there is technically no need to cache a bid if there's no ad server involved

Comment (karimMourra) [reply modules/videoModule/index.js:104]: it does feel like wasted work because `options.adXml` would be exactly the same as `options.prefetchedVastXml`

Comment (karimMourra) [reply modules/videoModule/index.js:104]: I think a publisher would (or should know to) only configure `config.cache` when using an ad server for secondary auction; i can't think of any other practical use case for caching

**Review by karimMourra — APPROVED** [review]
